### PR TITLE
Post component + Tree of posts

### PR DESCRIPTION
**Pull request by paolina92**
This PR is related to DEVAS-570 and DEVAS-476

### CODE REVIEW

File: assembl/static2/js/app/components/debate/thread/post.jsx [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    const totalSentimentsCount =
+      sentimentCounts.like + sentimentCounts.disagree + sentimentCounts.dontUnderstand + sentimentCounts.moreInfo;
+    return (
+      <div className="posts">
+        <div className="box">
+          <Row className="post-row">
+            <Col xs={12} md={11} className="post-left">
+              <ProfileLine userId={creator.userId} userName={creator.name} creationDate={creationDate} />
+              {/* TODO convert creationDate to "x months ago" with momentjs */}
+              <h3 className="dark-title-3">{parentId !== null ? <span>Rep. 1 :</span> : null}{subject}</h3>
```
Comment (ayyazdaniaryan) [inline assembl/static2/js/app/components/debate/thread/post.jsx]: `Rep. 1`?

File: assembl/static2/js/app/components/debate/thread/post.jsx [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+      body,
+      parentId,
+      indirectIdeaContentLinks,
+      mySentiment,
+      sentimentCounts,
+      creator,
+      creationDate
+    } = this.props;
+    let count = 1;
+    const totalSentimentsCount =
```
Comment (ayyazdaniaryan) [inline assembl/static2/js/app/components/debate/thread/post.jsx]: Should cache so only calculate once. So inefficient to re-calculate everytime the post renders, which would be a few times.

File: assembl/static2/js/app/components/debate/thread/post.jsx [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+  return { expanded: post.get('showResponses', SHOW_POST_RESPONSES) };
+});
+
+export const connectPostToState = connect(postMapStateToProps);
+
+export const PostFolded = ({ creator }) => {
+  return <Translate value="debate.thread.foldedPostLink" creatorName={creator.name} />;
+};
+
+export default class Post extends React.Component {
```
Comment (ayyazdaniaryan) [inline assembl/static2/js/app/components/debate/thread/post.jsx]: Until this component has more needs for lifecycle hooks etc, can it not become a functional component?

File: assembl/static2/js/app/constants.js [hunk 4]
```diff
@@ -0,0 +1 @@
+export const SHOW_POST_RESPONSES = true; // should we show all post responses while loading the posts thread?
```
Comment (ayyazdaniaryan) [inline assembl/static2/js/app/constants.js]: Cool, down to have all constants in one place.

File: assembl/static2/package.json [hunk 5]
```diff
@@ -72,9 +74,22 @@
...
     "react-tweet-embed": "^1.0.3",
+    "react-virtualized": "^9.8.0",
     "redux": "^3.6.0",
     "redux-thunk": "^2.1.0",
+    "reselect": "^3.0.1",
     "searchkit": "1.0.0-7",
     "url-join": "^2.0.0"
+  },
+  "jest": {
+    "//": "See https://facebook.github.io/jest/docs/webpack.html for more info",
```
Comment (ayyazdaniaryan) [inline assembl/static2/package.json]: Thank you for more instructions.

File: assembl/static2/js/app/selectors.js [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+
+export const localeSelector = (state) => {
+  return state.i18n.locale;
+};
+
+export const idSelector = (_, props) => {
+  return props.id;
+};
+
+export const postSelector = createSelector(postsByIdSelector, idSelector, (postsById, id) => {
```
Comment (ayyazdaniaryan) [inline assembl/static2/js/app/selectors.js]: Huh? `(postsById, id)`, why in brackets?

File: assembl/static2/js/app/reducers/postsReducer.js [hunk 7]
```diff
@@ -0,0 +1,14 @@
+import { Map } from 'immutable';
```
Comment (ayyazdaniaryan) [inline assembl/static2/js/app/reducers/postsReducer.js]: 👍

File: assembl/static2/js/app/pages/idea.jsx [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+export const transformPosts = (posts) => {
+  let postsByParent = Map();
+  posts.forEach((p) => {
+    const items = postsByParent.get(p.parentId, List());
+    postsByParent = postsByParent.set(p.parentId, items.push(p));
+  });
+
+  const getChildren = (id) => {
+    return postsByParent.get(id, List()).toJS().map((post) => {
+      return { ...post, children: getChildren(post.id) };
```
Comment (ayyazdaniaryan) [inline assembl/static2/js/app/pages/idea.jsx]: Perfect. 👍

File: assembl/static2/css/components/posts.scss [hunk 9]
```diff
@@ -79,6 +79,107 @@
...
+        overflow: hidden;
+    }
+
+    .post-left {
+        padding: 35px;
+    }
+
+    .post-right {
+        background-color: $first-color-light;
+        margin-bottom: -99999px!important;
```
Comment (paolina92) [inline assembl/static2/css/components/posts.scss]: Yes, you're right but I override a bootstrap class

File: assembl/static2/js/app/components/common/tree.jsx [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+import { AutoSizer, CellMeasurer, CellMeasurerCache, List, WindowScroller } from 'react-virtualized';
+
+/*
+  TODO: avoid globalList
+  TODO: InnerComponentFolded should be able to toggle the item
+*/
+
+let globalList;
+const cache = new CellMeasurerCache({
+  defaultHeight: 600,
```
Comment (paolina92) [inline assembl/static2/js/app/components/common/tree.jsx]: The height is re-calculate later. It's just a default height.

File: assembl/static2/js/app/components/common/tree.jsx [hunk 11]
```diff
@@ -0,0 +1,168 @@
+import React from 'react';
+import { AutoSizer, List } from 'react-virtualized';
```
Comment (swergas) [inline assembl/static2/js/app/components/common/tree.jsx]: What are the advantages and drawbacks of using this library compared to others?

File: assembl/static2/js/app/selectors.js [hunk 12]
```diff
@@ -0,0 +1,18 @@
+import { Map } from 'immutable';
+import { createSelector } from 'reselect';
```
Comment (swergas) [inline assembl/static2/js/app/selectors.js]: Same remark here about the add of libraries.

**Pull request merged**